### DISCUSSION TITLE
LilyGo modules output power

### DISCUSSION THREAD

**Discussion by tk5ep** [jgromes/RadioLib#903] [General]
Hi,
I wanted to test how much power the Lilygo T3 and T-beam modules are producing, as I do have a suspicious module.
I wrote a small script that uses radiolib FSK transmission, so that I have a steady carrier I can measure.

I set the power with setOutputPower(20)

When I measure the output power, On healthy modules I only get about 26 mW which is around +14 dBm.
On the suspicious one, 5mW what confirms that it is somewhere broken.

Looking at the Lilygo schematics, there is no PA_BOOST output pin, only one ANT output.
I didn't find any output power informations on Lilygos pages.  :-(

Is there a difference in output power handling when using FSK or LoRa with Radiolib ?

Any suggestion on how to increase the output power or can someone confirm my results ?

Thanks,

**Comment by jgromes** [comment]
I'm going to assume there is an SX127x in your module, since you didn't mention which transceiver this is about. SX127x has two possible RF output pins: PA_BOOST or RFO. You have to select the correct one, otherwise you will be transmitting into an unconneted pin, which is a quick way to destroy the transceiver. RadioLib assumes PA_BOOST is used by default, since that seems to be case on almost all off-the-shelf modules. You can select RFO by providing a second argument the output power method: `setOutputPower(int8_t power, bool useRfo)`

> Looking at the Lilygo schematics, there is no PA_BOOST output pin, only one ANT output.

I don't know what schematic you are referring to, does it show the SX127x transceiver? As mentioned above, and in the SX127x datasheet, it has either RFO or PA_BOOST. It has no ANT pin, so you might be looking at a pinout of just the TTGO module.

> Is there a difference in output power handling when using FSK or LoRa with Radiolib 

There is no difference there.

**Comment by tk5ep** [comment]
Hi, thanks for the fast and precise answer.
I was talking about Lilygo T-beam and T3 modules.
Being curious, I've just removed their sticker and the chip is a HPD14A V1.2. 

I've looked on their advertisement and here an extract related to the power :
+20 dBm - 100 mW constant RF output vs. V supply.
+14 dBm high efficiency PA.

They also say it is a SX1278.  ;-)

There is something I don't understand. +20dBm output and +14dBm PA ? 

This is the LoRa chip part of the complete diagram.
![T3-Lora-chip](https://github.com/jgromes/RadioLib/assets/1655173/aba3e612-f1c8-472b-bcab-b8e8ee6aa110)

As you can see, there is no pin choice for the amplifier output.
I didn't look on their datasheet yet.

OK about the power not being related to mode, so there is something weird.

Thanks,

**Reply by jgromes** [reply]
> This is the LoRa chip part of the complete diagram.

OK, let's see what the SX1278 datasheet has to say:

![Screenshot_77](https://github.com/jgromes/RadioLib/assets/3407192/c26ab7eb-d486-430e-b78d-38fa1904d9d8)

That's the pinout of the actual RF transceiver chip. In your case, either PA_BOOST or RFO_LF is connected to the "ANT" pin, probably through some impedance matching network. You need to figure out which in order to send the output signal to the correct pin.

> There is something I don't understand. +20dBm output and +14dBm PA ?

I don't understand that either. SX1278 has two PAs, one on the PA_BOOST pin (+2 to +17 dBm and a high power + 20 dBm mode), the other on the RFO (-1 to +14 dBm).  That's something the manufacturer should clarify. If they don't, why buy their stuff?

**Comment by tk5ep** [comment]
Hi
If made some more tests and researches.
1st, I checked with a known power source that my measures are +/- correct.
Then I confirmed the general low power outputs in testing a brand new T3 module. Same results : +14dBm power.

When trying to use the RFO output (setOutputPower(20,1), I do have far less output power using the +20dBm level and nothing with +10dBm. (at leat what I can measure)
Either this output pin is not used and I measure some power leaking inside the chip, or there is something weird with the power level handling.

These HPD14 modules are not described anywhere, no docs, no datasheet.
They look VERY similar to HopeRF RFM98 and NiceRF modules. They are probably the same with a metallic cover.
Without cover, I could have checked how the different SX1278 pins are wired on the board.

I'm not smart enough to check in the library how all these parameters (PA_BOOST, RegPaDac, RegOcp) are handled when trying to use +20bDM output.

Something else, is there a program example with which I could check the sensitivity of reception part ? Just to check if the RF path from connector to chip is correct ?

Thanks again,
Patrick.

**Reply by jgromes** [reply]
> Either this output pin is not used

It certainly seems that way.

> I'm not smart enough to check in the library how all these parameters (PA_BOOST, RegPaDac, RegOcp) are handled when trying to use +20bDM output.

You just have to `setOutputPower(20)`, the library will automatically enable high power mode on the PA_BOOST pin. The default OCP value is set for the default 10 dBm power, you can increase it to the maximmum by `setCurrentLimit(240)`.

> is there a program example with which I could check the sensitivity of reception part

Not sure what you mean, there is no specific example for sensitivity testing. You can try continuously reading RSSI in some loop: `getRSSI(false, true, 0)`

**Comment by tk5ep** [comment]
Something else, 
I looked on the signal with an SDR key and here the spectrum I get. Strong spurs around the signal about 15 dB lower and it looks like if it is modulated.
Indeed do I here a 2.1 kHz tone.
![lora-spectrum](https://github.com/jgromes/RadioLib/assets/1655173/7de9071a-ced7-4576-ab8c-f8d1fadbd861)

Do you have any explanation ?
Is there a better way then FSK to produce a continuous wave signal ?

Thanks,

**Reply by jgromes** [reply]
You haven't posted any code, so I don't know what exactly you are doing. In any case, you should be able to use `transmitDirect` to switch the module to direct transmission. In RadioLib this is used to emulate modes like RTTY.

**Comment by tk5ep** [comment]
OK Jan,

What I try to achieve, is to get a non modulated carrier at a level I want.
This, to measure the power and frequency.

I've added radio.setCurrentLimit(240), it did increased the output power to +17.5 dBm ! Still some dB missing, but it's better !
I then reduced the value to 100 to be safe, and the power is sill the same.
When I go under this current limit level, the output power gets limited.

I noticed that from +2-17 dBm, the power follows the setting.
Values 18 and 19 do produce the same as +10 dBm.
+20dBm produces the same as +17 dBm.
Looking at your library, I understand the values 18 & 19 are not handled and that in my case, the RADIOLIB_SX127X_PA_BOOST_ON is not activated. (by library or module)

I don't understand why I don't have a pure carrier but a modulated one.
FSK and RTTY shouldn't produce that.
It's not a big deal in my case, but I'm curious to understand.

Meanwhile, I managed to write a small script that displays the RSSI signal on the OLED screen.
It confirms that there is a problem in the RF path, the sensitivity is far lower than with  healthy module.  :-(

Here the code of the TX test I use with only the part concerning Radiolib.

```
int TXdelay   = 10000;      // TX delay in ms at each power
double TXfreq = 433.775;    // frequency in MHz
int TXpowerHi = 20;         // High power level in dBm (max 20)
int TXpowerLo = 10;         // Low power level in dBm

#include <RadioLib.h>
SX1278 radio = new Module(LORA_SS, LORA_DIO0, LORA_RST, LORA_DIO1);
RTTYClient rtty(&radio);

void setup() {
  radio.beginFSK();
  radio.setCurrentLimit(240);
  rtty.begin(TXfreq, 50, 45);
}

loop() {
	radio.setOutputPower(TXpowerHi);
	rtty(idle);
	delay(TXdelay);
	radio.setOutputPower(TXpowerLo);   
	rtty(idle);
	delay(TXdelay);
}
```

Sorry for the long post.
Thanks again for your kind support.

**Comment by tk5ep** [comment]
BTW how can I detect the chip version under the hood of the HPD14 module ?
SX1268 or SX1278 ?

**Reply by jgromes** [reply]
It's not a different "version", it's an entirely different chip. If you can communicate with it using the `SX1278` class, then it is an SX1278 chip. Otherwise you would have gotten -2 error (chip not found), because the SPI interface of SX126x is completely different to that of SX127x.

**Reply by tk5ep** [reply]
OK, thanks.
DIO1not being connected on the Lilygo boards, could this explain the RTTYclient problem ?
This 2400 Hz modulation is puzzling me.

I tried with the MorseClient, and I do get a better carrier.
But setting the output power during transmission doesn't change anything. Is there something to do to make this work ? 

Is the getFrequencyError() function working only in packet mode ? It returns 0 for now.

Thanks,

**Reply by jgromes** [reply]
> DIO1not being connected on the Lilygo boards, could this explain the RTTYclient problem ?

No, it's not used in that case.

> This 2400 Hz modulation is puzzling me.

What is `rtty(idle);` doing? Did you mean `rtty.idle();` or is that some custom function you have not shown? In any case, I don't understand why are you using RTTY at all for this, you can call `transmitDirect()` instead. You might try switching to OOK to get rid of the FSK modulation.

> Is the getFrequencyError() function working only in packet mode ? It returns 0 for now.

That information (including everything else) is in the [documentation](https://jgromes.github.io/RadioLib/class_s_x127x.html#af6aa854a2668d70f4d3a374a49440362).

**Reply by tk5ep** [reply]
Hi, I finally managed to make what I wanted using the AFSKClient.
Just one thing. The first time I call it, I get this error :
` E (8670) ledc: ledc_get_duty(740): LEDC is not initialized`

**Reply by jgromes** [reply]
>  The first time I call it, I get this error

Which exact method are you calling? `AFSKClinet::tone()`? And if so, which version of ESP-IDF are you using?

**Comment by tk5ep** [comment]
I'm using a T-beam board, so ESP32.
VSC and PlatformIO.

Here part of my code :

int TXpowerHi = 20;
int TXpowerLo = 10;
SX1278 radio = new Module(LORA_SS, LORA_DIO0, LORA_RST, LORA_DIO1);
AFSKClient audio(&radio,32);

void setup() {
  int state = radio.beginFSK();
  state = radio.setFrequency(TXfreq);
  state = audio.begin();
}

void loop() {
    int state = radio.setOutputPower(TXpowerHi);
    audio.tone(1500);
    radio.receiveDirect();
    audio.tone(1000);
    state = radio.setOutputPower(TXpowerLo);
    radio.transmitDirect();
}


Le mer. 20 déc. 2023 à 10:23, Jan Gromeš ***@***.***> a
écrit :

> The first time I call it, I get this error
>
> Which exact method are you calling? AFSKClinet::tone()? And if so, which
> version of ESP-IDF are you using?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jgromes/RadioLib/discussions/903#discussioncomment-7906171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMUDBKNNFAVC57NWBTZZ73YKKVCTAVCNFSM6AAAAABAYGBYKCVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TSMBWGE3TC>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Patrick Egloff
email : ***@***.***
Web page : http://www.egloff.eu

**Reply by jgromes** [reply]
> I'm using a T-beam board, so ESP32.
> VSC and PlatformIO.

Yes, but which *version* of the ESP-IDF framework? In version 5.0.0, ldec API changed. RadioLib should be able to cover that change, so the error ` E (8670) ledc: ledc_get_duty(740): LEDC is not initialized` you got is strange.

Also:

```c++
void loop() {
    int state = radio.setOutputPower(TXpowerHi);
    audio.tone(1500);
    radio.receiveDirect();
    audio.tone(1000);
    state = radio.setOutputPower(TXpowerLo);
    radio.transmitDirect();
}
```

I don't understand what exactly you are trying to do there. You set Tx power, start transmitting 1500 Hz tone, but then immediately switch to Rx, then start transmitting 100 Hz tone, try to change the output power (which I don't think works, and if it does then it really shouldn't), then re-enable direct transmission and then do it all over again ... ?

**Reply by tk5ep** [reply]
Hi,
I think I have version 4.4.5. (that's at least what a call to ESP.getSdkVersion() returns)

> I don't understand what exactly you are trying to do there. You set Tx power, start transmitting 1500 Hz tone, but then immediately switch to Rx, then start transmitting 100 Hz tone, try to change the output power (which I don't think works, and if it does then it really shouldn't), then re-enable direct transmission and then do it all over again ... ?

Yes, this is EXACTLY what I try to achieve, and so far it works.
I did remove the delay instructions in the loop in my post. Each power level lasts 10s.

Without switching to RX, the setOutputPower did'nt work, returning error -16.

If this is not the right way to do it, how should I do this ?

Thanks,

**Reply by jgromes** [reply]
> I did remove the delay instructions

Oh, so that's why it did not make sense to me. Assuming there are delays (somewhere), it does.

> Without switching to RX, the setOutputPower did'nt work, returning error -16.
> If this is not the right way to do it, how should I do this ?

By explicitly setting the mode to standby by  `radio.standby()`.

**Reply by tk5ep** [reply]
OK thanks,

It works now perfectly as expected.
